url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/Datatypes.java

```diff
@@ -0,0 +1,890 @@
+	 * <tr><td><tt>Cc</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Cc</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Cf</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Cf</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Co</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Co</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Cn</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Cn</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>C</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">C</a> category from Unicode 3.1</td></tr>
+	 * </table>
+	 * <p>Loaded automata are cached in memory.
+	 *
+	 * @param name name of automaton
+	 * @return automaton
+	 */
+	public static DefaultAutomaton get(String name) {
+		DefaultAutomaton a = automata.get(name);
+		if (a == null) {
+			a = load(name);
+			automata.put(name, a);
+		}
+		return a;
+	}
+
+	/**
+	 * Checks whether the given string is the name of a Unicode block (see {@link #get(String)}).
+	 */
+	public static boolean isUnicodeBlockName(String name) {
+		return unicodeblock_names.contains(name);
+	}
+
+	/**
+	 * Checks whether the given string is the name of a Unicode category (see {@link #get(String)}).
+	 */
+	public static boolean isUnicodeCategoryName(String name) {
+		return unicodecategory_names.contains(name);
+	}
+
+	/**
+	 * Checks whether the given string is the name of an XML / XML Schema automaton (see {@link #get(String)}).
+	 */
+	public static boolean isXMLName(String name) {
+		return xml_names.contains(name);
+	}
+
+	/**
+	 * Checks whether a given automaton is available.
+	 *
+	 * @param name automaton name
+	 * @return true if the automaton is available
+	 */
+	public static boolean exists(String name) {
+		try {
+			Datatypes.class.getClassLoader().getResource(name + ".aut").openStream().close();
+		} catch (IOException e) {
+			return false;
+		}
+		return true;
+	}
+
+	private static DefaultAutomaton load(String name) {
+		try {
+			URL url = Datatypes.class.getClassLoader().getResource(name + ".aut");
+			return (DefaultAutomaton) new ObjectInputStream(url.openStream()).readObject();
+		} catch (IOException e) {
+			e.printStackTrace();
+			return null;
+		} catch (ClassNotFoundException e) {
+			e.printStackTrace();
+			return null;
+		}
+	}
+
+	private static void store(String name, DefaultAutomaton a) {
+		String dir = System.getProperty("dk.brics.automaton.datatypes");
+		if (dir == null)
+			dir = "build";
+		try {
+			ObjectOutputStream o = new ObjectOutputStream((new FileOutputStream(dir + "/" + name + ".aut")));
+			o.writeObject(a);
+			o.flush();
+			o.close();
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private static void buildAll() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 146 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 502-695, lines 803-852 into separate methods.